crates/sargon/src/system/sargon_os/transactions/sargon_os_transaction_analysis.rs

```diff
@@ -0,0 +1,759 @@
+                        .reserved_instructions
+                        .iter()
+                        .map(|i| i.to_string())
+                        .collect(),
+                },
+            );
+        }
+
+        Ok(TransactionToReview {
+            transaction_manifest,
+            execution_summary,
+        })
+    }
+
+    async fn get_transaction_preview(
+        &self,
+        gateway_client: GatewayClient,
+        manifest: TransactionManifest,
+        network_id: NetworkID,
+        message: Message,
+        nonce: Nonce,
+        notary_public_key: PublicKey,
+    ) -> Result<TransactionPreviewResponse> {
+        let signer_public_keys = self
+            .extract_transaction_signer_public_keys(manifest.summary())
+            .await?;
+        let epoch = gateway_client.current_epoch().await?;
+        let header = TransactionHeader::new(
+            network_id,
+            epoch,
+            Epoch::window_end_from_start(epoch),
+            nonce,
+            notary_public_key,
+            signer_public_keys.is_empty(),
+            0,
+        );
+        let intent = TransactionIntent::new(header, manifest, message)?;
+        let request = TransactionPreviewRequest::new(
+            intent,
+            signer_public_keys,
+            TransactionPreviewRequestFlags::default(),
+        );
+        let response = gateway_client.transaction_preview(request).await?;
+        if response.receipt.status != TransactionReceiptStatus::Succeeded {
+            return Err(Self::map_failed_transaction_preview(response));
+        };
+        Ok(response)
+    }
+
+    fn map_failed_transaction_preview(
+        response: TransactionPreviewResponse,
+    ) -> CommonError {
+        let message = response
+            .receipt
+            .error_message
+            .unwrap_or_else(|| "Unknown reason".to_string());
+
+        // Quite rudimentary, but it is not worth making something smarter,
+        // as the GW will provide in the future strongly typed errors
+        let is_failure_due_to_deposit_rules = message
+            .contains(GW_ERR_ACCOUNT_DEPOSIT_DISALLOWED)
+            || message.contains(GW_ERR_NOT_ALL_COULD_BE_DEPOSITED);
+
+        if is_failure_due_to_deposit_rules {
+            CommonError::OneOfReceivingAccountsDoesNotAllowDeposits
+        } else {
+            CommonError::FailedTransactionPreview {
+                error_message: message,
+            }
+        }
+    }
+
+    async fn extract_transaction_signer_public_keys(
+        &self,
+        manifest_summary: ManifestSummary,
+    ) -> Result<IndexSet<PublicKey>> {
+        let signer_entities = self
+            .extract_transaction_signer_entities(manifest_summary.clone())
+            .await?;
+        Ok(self
+            .extract_transaction_signers_factor_instances(signer_entities)
+            .iter()
+            .map(|fi| fi.public_key)
+            .collect())
+    }
+
+    async fn extract_transaction_signer_entities(
```

Comment:
@sergiupuhalschi-rdx @micbakos-rdx important, this method should be moved into the `SignatureCollector` when @micbakos-rdx merges #235 , see here is the near identical code in in that PR (wait for it... GH was slow expanding the link for me): https://github.com/radixdlt/sargon/pull/235/files#diff-2d257ce713ef9ec886a81af3229aae3b41a3e312b8d8ba2a35eccb5ef145bb13R29

when in doubt prefer @micbakos-rdx s implementation. Ensure to test hidden entities... @GhenadieVP should Sargon fail if any of the entities "requiring auth" is hidden?

Reply:
> should Sargon fail if any of the entities "requiring auth" is hidden

No, the respective entity would be considered to be a foreign entity.